srml/contract/src/account_db.rs

```diff
@@ -43,8 +44,44 @@ impl<T: Trait> Default for ChangeEntry<T> {
 
 pub type ChangeSet<T> = BTreeMap<<T as system::Trait>::AccountId, ChangeEntry<T>>;
 
+#[derive(Clone, Default)]
+pub struct AccountKeySpaceMapping<A: Ord> {
+	to_account: BTreeMap<KeySpace, A>,
+	to_key: BTreeMap<A, KeySpace>,
+}
+
+impl<A: Clone + Ord> AccountKeySpaceMapping<A> {
+
+	pub fn new() -> Self {
+		AccountKeySpaceMapping {
+			to_account: BTreeMap::new(),
+			to_key: BTreeMap::new(),
+		}
+	}
+
+	pub fn insert(&mut self, account: A, ks: KeySpace) {
+		self.to_account.insert(ks.clone(), account.clone());
+		self.to_key.insert(account, ks);
+	}
+	pub fn get_keyspace(&self, account: &A) -> Option<&KeySpace> {
+		self.to_key.get(account)
+	}
+	pub fn get_account(&self, ks: &KeySpace) -> Option<&A> {
+		self.to_account.get(ks)
+	}
+
+}
+
 pub trait AccountDb<T: Trait> {
-	fn get_storage(&self, account: &T::AccountId, location: &[u8]) -> Option<Vec<u8>>;
+	fn get_subtrie(&self, account: &T::AccountId) -> Option<SubTrie>;
+	fn get_or_create_keyspace(&self, account: &T::AccountId) -> KeySpace;
+	// this function is borderline (impl specific to overlaydb). An alternate
+	// design is to use rc internally (instead of optional cache) for 
+	// overlaydb. It avoid querying the cache through a recursive call,
+	// but make `get_subtrie` and `get_or_create_keyspace` recursive call
+	// awkward : multiple query or insert (unless we specify top overlaydb instance).
+	fn get_account_from_reg_ks(&self, ks: &KeySpace) -> Option<T::AccountId>;
```

Comment:
This is used for get_storage of overlaydb. Am I wrong if I say that `AccountDB` trait is not used out of this module? I believe it could be seen as a technical trait for the overlay+direct linked list of layers.
Switching to Rc could remove it.

Reply:
> Am I wrong if I say that AccountDB trait is not used out of this module?

When I say "within this module" I mean a Rust module or ".rs" file. So yeah, the `AccountDb` trait is **used** outside of this module. However, this trait is not used outside of this crate.

Reply:
Oups, I indeed misunderstood, this method is bad design (as written in the code comment), but the way the linked account overlay list run I cannot really think of an alternative (except the Rc design).